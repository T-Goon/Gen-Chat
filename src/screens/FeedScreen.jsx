import React, { useContext } from "react";
import { View, Text, StyleSheet, FlatList, Pressable } from "react-native";
import { FontAwesomeIcon } from '@fortawesome/react-native-fontawesome';
import { faCog, faPlusCircle } from '@fortawesome/free-solid-svg-icons';

import FeedItem from "../components/FeedItem";

import { Context } from "../Context";

const FeedScreen = ({ navigation }) => {
    const { messages } = useContext(Context);

    const renderItem = ({ item }) => {
        const json = JSON.parse(item);
        return (
            <FeedItem username={json.username} message={json.message} />
        );
    };

    return (
        <View style={styles.container}>
            <View style={styles.headerContainer}>
                <Text style={styles.header}>Feed</Text>
                <Pressable testID='settings-button' style={styles.settingsButton} onPress={() => { navigation.navigate('Settings'); }}>
                    <FontAwesomeIcon icon={faCog} size={28} />
                </Pressable>
            </View>
            <FlatList
                style={styles.list}
                data={messages}
                renderItem={renderItem}
                keyExtractor={(item, index) => index} />
            <Pressable testID='add-post-button' style={styles.postButton} onPress={() => { navigation.navigate('Send Post'); }}>
                <FontAwesomeIcon icon={faPlusCircle} color='hsla(0, 0%, 0%,.5)' size={50} />
            </Pressable>
        </View>
    );
};

const styles = StyleSheet.create({
    container: {
        flex: 1,
        marginTop: 39,
        marginBottom: 20,
        marginHorizontal: 15
    },
    headerContainer: {
        flexDirection: 'row',
        alignItems: 'center'
    },
    header: {
        fontSize: 30,
        fontWeight: 'bold'
    },
    settingsButton: {
        flex: 1,
        flexDirection: 'row-reverse'
    },
    list: {
    },
    postButton: {
        position: 'absolute',
        bottom: 0,
        right: 0
    }
});

export default FeedScreen;