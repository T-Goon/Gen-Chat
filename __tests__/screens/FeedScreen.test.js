import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';

import FeedScreen from '../../src/screens/FeedScreen';

jest.mock('../../src/Context', () => {
    const originalModule = jest.requireActual('../../src/Context');
    const { createContext } = jest.requireActual('react');

    const initialValue = {
        messages: [JSON.stringify({ username: 'testUser', message: 'testMessage' })]
    };

    return {
        __esModule: true,
        ...originalModule,
        Context: createContext(initialValue)
    };
});


describe('<FeedScreen />', () => {

    it('has i child', async () => {
        const tree = render(<FeedScreen />).toJSON();

        expect(tree.children.length).toBe(3);
    });

    it('renders correctly', async () => {
        const tree = render(<FeedScreen />).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('Go To Settings', async () => {
        const navigation = { navigate: jest.fn() };
        const feedScreen = render(<FeedScreen navigation={navigation} />);

        await fireEvent(feedScreen.getByTestId('settings-button'), 'press');
        await expect(navigation.navigate).toHaveBeenCalledWith('Settings');
    });

    it('Go To Add Post', async () => {
        const navigation = { navigate: jest.fn() };
        const feedScreen = render(<FeedScreen navigation={navigation} />);

        await fireEvent(feedScreen.getByTestId('add-post-button'), 'press');
        await expect(navigation.navigate).toHaveBeenCalledWith('Send Post');
    });
});