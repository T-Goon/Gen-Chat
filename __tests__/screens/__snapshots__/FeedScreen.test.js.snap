// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FeedScreen /> renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "marginBottom": 20,
      "marginHorizontal": 15,
      "marginTop": 39,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 30,
          "fontWeight": "bold",
        }
      }
    >
      Feed
    </Text>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "flex": 1,
          "flexDirection": "row-reverse",
        }
      }
      testID="settings-button"
    >
      <RNSVGSvgView
        align="xMidYMid"
        bbHeight={28}
        bbWidth={28}
        fill="#000"
        focusable={false}
        height={28}
        meetOrSlice={0}
        minX={0}
        minY={0}
        secondaryFill={null}
        secondaryOpacity={0.4}
        size={28}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "borderWidth": 0,
            },
            Object {},
            Object {
              "flex": 0,
              "height": 28,
              "width": 28,
            },
          ]
        }
        vbHeight={512}
        vbWidth={512}
        width={28}
      >
        <RNSVGGroup
          fill={4278190080}
          propList={
            Array [
              "fill",
            ]
          }
        >
          <RNSVGPath
            d="M487.4 315.7l-42.6-24.6c4.3-23.2 4.3-47 0-70.2l42.6-24.6c4.9-2.8 7.1-8.6 5.5-14-11.1-35.6-30-67.8-54.7-94.6-3.8-4.1-10-5.1-14.8-2.3L380.8 110c-17.9-15.4-38.5-27.3-60.8-35.1V25.8c0-5.6-3.9-10.5-9.4-11.7-36.7-8.2-74.3-7.8-109.2 0-5.5 1.2-9.4 6.1-9.4 11.7V75c-22.2 7.9-42.8 19.8-60.8 35.1L88.7 85.5c-4.9-2.8-11-1.9-14.8 2.3-24.7 26.7-43.6 58.9-54.7 94.6-1.7 5.4.6 11.2 5.5 14L67.3 221c-4.3 23.2-4.3 47 0 70.2l-42.6 24.6c-4.9 2.8-7.1 8.6-5.5 14 11.1 35.6 30 67.8 54.7 94.6 3.8 4.1 10 5.1 14.8 2.3l42.6-24.6c17.9 15.4 38.5 27.3 60.8 35.1v49.2c0 5.6 3.9 10.5 9.4 11.7 36.7 8.2 74.3 7.8 109.2 0 5.5-1.2 9.4-6.1 9.4-11.7v-49.2c22.2-7.9 42.8-19.8 60.8-35.1l42.6 24.6c4.9 2.8 11 1.9 14.8-2.3 24.7-26.7 43.6-58.9 54.7-94.6 1.5-5.5-.7-11.3-5.6-14.1zM256 336c-44.1 0-80-35.9-80-80s35.9-80 80-80 80 35.9 80 80-35.9 80-80 80z"
            fill={4278190080}
            fillOpacity={1}
            propList={
              Array [
                "fill",
                "fillOpacity",
              ]
            }
          />
        </RNSVGGroup>
      </RNSVGSvgView>
    </View>
  </View>
  <RCTScrollView
    data={
      Array [
        "{\\"username\\":\\"testUser\\",\\"message\\":\\"testMessage\\"}",
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={Object {}}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "backgroundColor": "hsl(230, 100%, 80%)",
              "borderRadius": 10,
              "marginVertical": 5,
              "padding": 5,
            }
          }
        >
          <Text
            style={
              Object {
                "fontSize": 20,
                "fontWeight": "bold",
              }
            }
          >
            testUser:
          </Text>
          <Text
            style={
              Object {
                "paddingLeft": 20,
              }
            }
          >
            testMessage
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "bottom": 0,
        "position": "absolute",
        "right": 0,
      }
    }
    testID="add-post-button"
  >
    <RNSVGSvgView
      align="xMidYMid"
      bbHeight={50}
      bbWidth={50}
      fill="hsla(0, 0%, 0%,.5)"
      focusable={false}
      height={50}
      meetOrSlice={0}
      minX={0}
      minY={0}
      secondaryFill={null}
      secondaryOpacity={0.4}
      size={50}
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          Object {},
          Object {
            "flex": 0,
            "height": 50,
            "width": 50,
          },
        ]
      }
      vbHeight={512}
      vbWidth={512}
      width={50}
    >
      <RNSVGGroup
        fill={2147483648}
        propList={
          Array [
            "fill",
          ]
        }
      >
        <RNSVGPath
          d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm144 276c0 6.6-5.4 12-12 12h-92v92c0 6.6-5.4 12-12 12h-56c-6.6 0-12-5.4-12-12v-92h-92c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h92v-92c0-6.6 5.4-12 12-12h56c6.6 0 12 5.4 12 12v92h92c6.6 0 12 5.4 12 12v56z"
          fill={2147483648}
          fillOpacity={1}
          propList={
            Array [
              "fill",
              "fillOpacity",
            ]
          }
        />
      </RNSVGGroup>
    </RNSVGSvgView>
  </View>
</View>
`;
